src/serde.rs

```diff
@@ -0,0 +1,87 @@
+use serde::{ser::Serialize, Deserialize};
+
+use crate::{NodeId, NodeRef, Tree};
+
+#[derive(Debug)]
+struct SerNode<'a, T> {
+    value: &'a T,
+    children: Vec<SerNode<'a, T>>,
+}
+
+impl<'a, T> From<NodeRef<'a, T>> for SerNode<'a, T> {
+    fn from(node: NodeRef<'a, T>) -> Self {
+        let value: &T = node.value();
+        let children: Vec<SerNode<'a, T>> = node.children().map(SerNode::<'a, T>::from).collect();
+        SerNode { value, children }
+    }
+}
+
+impl<'a, T: Serialize> Serialize for SerNode<'a, T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        (&self.value, &self.children).serialize(serializer)
+    }
+}
+
+impl<T: Serialize> Serialize for Tree<T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        SerNode::from(self.root()).serialize(serializer)
+    }
+}
+
+#[derive(Debug)]
+struct DeserNode<T> {
+    value: T,
+    children: Vec<DeserNode<T>>,
+}
+
+impl<T> DeserNode<T> {
+    fn into_tree_node(self, tree: &mut Tree<T>, parent: NodeId) -> NodeId {
+        let mut parent = tree.get_mut(parent).unwrap();
+        let node = parent.append(self.value).id();
+
+        for child in self.children {
+            child.into_tree_node(tree, node);
+        }
```

Comment:
I don't think we should change this, but we should document that the implementations for (de)serialization are recursive and require stack memory proportional to the tree depth.

Reply:
In addition, this new piece of documentation should be written as a module level doc comment

Reply:
Fixed